modules/shapeless/src/main/scala/io/parsek/shapeless/HListDecoder.scala

```diff
@@ -0,0 +1,78 @@
+package io.parsek.shapeless
+
+import io.parsek.Decoder.Result
+import io.parsek.PValue.{PMap, PNull}
+import io.parsek.instances.EncoderInstances
+import io.parsek.{Decoder, Encoder, PValue}
+import shapeless.labelled.FieldType
+import shapeless.{::, Default, HList, HNil, LabelledGeneric, Lazy, Witness}
+
+trait HListDecoder {
+
+  trait DecoderWithDefaults[T] {
+    def apply(v: PValue, defaults: Map[Symbol, PValue]): Result[T]
+  }
+
+  implicit def hlistObjectDecoderWithDefaults[K <: Symbol, H, T <: HList](
+     implicit
+     witness: Witness.Aux[K],
+     hDecoder: Lazy[DecoderWithDefaults[H]],
+     tDecoder: Lazy[DecoderWithDefaults[T]]
+    ): DecoderWithDefaults[FieldType[K, H] :: T] =
+    new DecoderWithDefaults[FieldType[K, H] :: T] {
+      override def apply(v: PValue, defaults: Map[Symbol, PValue]): Result[FieldType[K, H] :: T] = {
+        val headResult: Decoder.Result[FieldType[K, H]] = {
+          v match {
+            case p@PMap(valuesMap) =>
+              val fieldName = witness.value
+              valuesMap
+                .get(fieldName)
+                .orElse(defaults.get(fieldName))
```

Comment:
In case if default value for field is not defined hDecoder could now how to work with non existing data. For example  Decoder[Option[_]] could return None